src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,848 @@
+        this.array = array;
+        this._currentGroupIndex = -groupSize;
+        this._groupSize = groupSize;
+    }
+    /**
+     * Determines if the values of specified items in a group array are equal
+     * @private
+     * @param   {GroupArray} groupArray1 First group array
+     * @param   {GroupArray} groupNum1   Item within group array
+     * @param   {GroupArray} groupArray2 Second group array
+     * @param   {GroupArray} groupNum2   Item within 2nd group array
+     * @return {boolean}  true when values are equal
+     */
+    GroupArray.prototype.isGroupValuesEqual = function (groupNum1, groupArray2, groupNum2) {
+        var groupNumber;
+        var groupSize = this.groupSize();
+        for (groupNumber = 0; groupNumber < groupSize; groupNumber++) {
+            if (!_.isEqual(this.getGroupValue(groupNum1, groupNumber), groupArray2.getGroupValue(groupNum2, groupNumber))) {
+                return false;
+            }
+        }
+        return true;
+    };
+    GroupArray.prototype.nextGroupIndex = function () {
+        if (this._currentGroupIndex < this.array.length - this._groupSize) {
+            this._currentGroupIndex += this._groupSize;
+        } else {
+            this._currentGroupIndex = -this._groupSize;
+            return false;
+        }
+        return this._currentGroupIndex;
+    };
+    GroupArray.prototype.prevGroupIndex = function () {
+        if (this._currentGroupIndex - this._groupSize > -1) {
+            this._currentGroupIndex -= this._groupSize;
+        } else {
+            this._currentGroupIndex = -this._groupSize;
+            return false;
+        }
+        return this._currentGroupIndex;
+    };
+    GroupArray.prototype.setCurrentGroup = function (groupNumber) {this._currentGroupIndex = groupNumber * this._groupSize; };
+    GroupArray.prototype.getGroupIndex = function (groupNumber) { return this._groupSize * groupNumber; };
+    GroupArray.prototype.getGroupValue = function (groupNumber, valueIndexWithinGroup) {return this.array[(this._groupSize * groupNumber) + valueIndexWithinGroup]; };
+    GroupArray.prototype.currentGroupIndex = function () { return this._currentGroupIndex; };
+    GroupArray.prototype.currentGroupNumber = function () { return this._currentGroupIndex / this._groupSize; };
+    GroupArray.prototype.groupSize = function () { return this._groupSize; };
+    GroupArray.prototype.itemCount = function () { return this.array.length / this._groupSize; };
+
+    /**
+     * Enhances array with functions which facilitate managing the array contents
+     * by groups of items.
+     * This is useful for both performance and memory consumption to store the indexes
+     * of the match result beginning and ending locations.
+     * @private
+     * @param {Array} array The array to enhance
+     * @param {number} groupSize The number of indices that belong to a group
+     * @return {GroupArray} Enhanced Array
+     */
+    function _makeGroupArray(array, groupSize) {
```

Comment:
This is factory function pattern.  It is usually considered a good thing, but I will change if you wish.